### PR TITLE
HIVE-27428: CTAS fails with SemanticException when join subquery has complex type column and false filter predicate

### PR DESCRIPTION
**Pull request by kasakrisz**
### What changes were proposed in this pull request?
Create AST representation of `NULL` literals of RexLiterals with complex types in `ASTConverter`

### Why are the changes needed?
Such literals was converted to `NULL` without type information however CTAS statements require the exact type of all columns in the resulting schema of the query.

### Does this PR introduce _any_ user-facing change?
Yes. Such CTAS statements will be executed without error. See jira for example.

### How was this patch tested?
```
mvn test -Dtest.output.overwrite -Dtest=TestMiniLlapLocalCliDriver -Dqfile=vector_groupby_grouping_sets_grouping.q,vector_ptf_part_simple.q,min_structvalue.q,vector_coalesce.q,analyze_npe.q,empty_result_ctas.q -pl itests/qtest -Pitests
```

### CODE REVIEW

**Review by amansinha100 — APPROVED** [review]
Changes LGTM and simplifies the handling for null literals. 
+1

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/translator/ASTConverter.java [hunk 1]
```diff
@@ -756,17 +737,8 @@ public ASTNode visitInputRef(RexInputRef inputRef) {
...
-          return ASTBuilder.literal(literal);
-        }
-        nullLiteralMap.put(literal, true);
-        RexNode r = rexBuilder.makeAbstractCast(literal.getType(), literal);
-
-        return r.accept(this);
+      if (RexUtil.isNull(literal) && literal.getType().getSqlTypeName() != SqlTypeName.NULL) {
+        return createNullField(literal.getType());
       }
       return ASTBuilder.literal(literal);
```
Comment (zabetak) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/translator/ASTConverter.java:743]: Maybe in a follow up we could add some unit tests for this visitor and in particular for this method so that we can test all supported null/not null types. Not strictly needed as part of this change.

File: ql/src/test/results/clientpositive/llap/analyze_npe.q.out [hunk 2]
```diff
@@ -159,7 +159,7 @@ STAGE PLANS:
           Filter Operator
             predicate: c1 is null (type: boolean)
             Select Operator
-              expressions: null (type: void)
+              expressions: array(null) (type: array<string>)
```
Comment (zabetak) [inline ql/src/test/results/clientpositive/llap/analyze_npe.q.out:162]: I understand the changes in this file but in some databases such nuances (`null` vs `array(null)`) lead to different results.
I run the following on Postgres and confirmed that plain null and an array with nulls is not the same thing.
```
create table tbl_arr(ca INT ARRAY);
insert into tbl_arr values ('{1}');
insert into tbl_arr values ('{2,3}');
insert into tbl_arr values (null);
insert into tbl_arr values ('{null}');

select * from tbl_arr;
select * from tbl_arr where ca is null;
select * from tbl_arr where ca[1] is null;
```
Can you please check what Hive does/should do and if the results are different after your changes. This file has just `EXPLAIN` statements so not easy to tell.

**Review by zabetak — COMMENTED** [review]
LGTM. If there is no unintended change in the result set of queries due the difference `null` vs `complex(null)` I am +1 on merging this.
